components/service/nimbus/src/main/java/mozilla/components/service/nimbus/Nimbus.kt

```diff
@@ -161,15 +158,85 @@ class Nimbus : NimbusApi {
 
     override fun updateExperiments() {
         if (!isInitialized) return
-        nimbus.updateExperiments()
-        onExperimentUpdated?.invoke(nimbus.getActiveExperiments())
+        // We need to launch this off of the main thread because it involves a network request
+        // within the nimbus-sdk.
+        scope.launch {
+            if (tryUpdateExperiments()) {
+                tryGetActiveExperiments()?.let {
+                    recordExperimentTelemetry(it)
+                    onExperimentUpdated?.invoke(it)
+                }
+            }
+        }
     }
 
     override fun optOut(experimentId: String) {
         if (!isInitialized) return
         nimbus.optOut(experimentId)
     }
 
+    // Wraps the `Nimbus.updateExperiments()` in a try/catch since the underlying nimbus-sdk
+    // makes a network request that might throw an error.  This also wraps the call in
+    // a simple retry mechanism before giving up.
+    // Returns true if the update completed without errors, false otherwise.
+    // Note: a return of "true" doesn't mean that there was any new experiments, just that there
+    // were no errors.
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun tryUpdateExperiments(): Boolean {
+        var retries = 0
+        while (retries++ < MAX_RETRY_COUNT) {
+            // Get experiments
+            try {
+                nimbus.updateExperiments()
+                return true
+            } catch (e: ErrorException.RequestError) {
+                Log.log(Log.Priority.ERROR,
+                    LOG_TAG,
+                    message = "Error fetching experiments from endpoint: $e"
+                )
+            } catch (e: ErrorException.InvalidExperimentResponse) {
+                Log.log(Log.Priority.ERROR,
+                    LOG_TAG,
+                    message = "Invalid experiment response: $e"
+                )
+            }
+
+            Thread.sleep(RETRY_DELAY_MS)
+        }
+
+        return false
+    }
+
+    // Wraps the `Nimbus.getActiveExperiments()` in a try/catch since the underlying nimbus-sdk
+    // call could make a network request that might throw an error.  This also wraps the call in
+    // a simple retry mechanism before giving up.
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun tryGetActiveExperiments(): List<EnrolledExperiment>? {
+        var retries = 0
+        var activeExperiments: List<EnrolledExperiment>? = null
+        while (retries++ < MAX_RETRY_COUNT) {
+            // Get experiments
+            try {
+                activeExperiments = nimbus.getActiveExperiments()
+                break
+            } catch (e: ErrorException.RequestError) {
+                Log.log(Log.Priority.ERROR,
+                    LOG_TAG,
+                    message = "Error fetching experiments from endpoint: $e"
+                )
```

Comment:
nit: The logging API also allows you to create a `Logger(tag)` in your class. Then calling this here is a bit nicer and you only need `logger.error(...)`.

nit 2: "Error" seems to be a pretty high level for something that happens frequently and is expected (no network?)?

Reply:
I can tone it down to an info or debug.